### PR TITLE
errors: add ability to send stack traces to sentry

### PR DESCRIPTION
**Pull request by sergiusens**
Adds the ability to send reports, asking for confirmation before doing
so. This feature is only enabled for the deb as raven is not provided in
`xenial` and most of the user base will be using the snap.

Closes: #1918

- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [x] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh unit`?

-----

### CODE REVIEW

File: snapcraft/cli/_errors.py [hunk 1]
```diff
@@ -20,6 +20,19 @@
...
+# raven is not available on 16.04
+try:
+    from raven import Client as RavenClient
+    from raven.transport import RequestsHTTPTransport
+except ImportError:
+    RavenClient = None
+
+_SEND_TO_SENTRY = (
+    'Sorry, Snapcraft had an internal error.\n'
+    'To help us improve, would you like to send error data?'
```
Comment (kalikiana) [inline snapcraft/cli/_errors.py]: This message raises a couple of questions in my head:

- What data is this going to send?
- Will this create a bug report a la apport or ubuntu-bug or something else?

That is to say, perhaps the message can be more helpful in telling the user what is going to happen with regard to privacy concerns and what Snapcraft is going to do.

File: snapcraft/cli/_errors.py [hunk 2]
```diff
@@ -46,6 +59,9 @@ def exception_handler(exception_type, exception, exception_traceback, *,
         traceback.print_exception(
             exception_type, exception, exception_traceback)
 
+    if not is_snapcraft_error and RavenClient is not None:
+        _submit_trace(exception)
```
Comment (kalikiana) [inline snapcraft/cli/_errors.py]: Can we have a test case for this? It seems like there's three cases here at least that we'll care about.

1. Trace prompt will be triggered
2. Non-interactive build skips past the prompt (rather than timing out)
3. No trace available

File: snapcraft/cli/_errors.py [hunk 3]
```diff
@@ -55,3 +71,15 @@ def exception_handler(exception_type, exception, exception_traceback, *,
             echo.error(str(exception))
 
     sys.exit(exit_code)
+
+
+def _submit_trace(exception):
+    client = RavenClient(
+        'https://b0fef3e0ced2443c92143ae0d038b0a4:'
+        'b7c67d7fa4ee46caae12b29a80594c54@sentry.io/277754',
```
Comment (elopio) [inline snapcraft/cli/_errors.py]: Curious, how can we prevent to be spammed with evil messages to our sentry?

Comment (sergiusens) [reply snapcraft/cli/_errors.py]: I don't think we can.

File: requirements.txt [hunk 4]
```diff
@@ -17,6 +17,7 @@ pymacaroons==0.9.2; sys_platform != 'win32'
 pymacaroons==0.10.0; sys_platform == 'win32'
 pymacaroons-pynacl==0.9.3
 pysha3==1.0b1; python_version < '3.6'
+raven==6.5.0
```
Comment (elopio) [inline requirements.txt]: I think it would be good to explain somewhere, maybe in the description of this pr, why raven is not a requirement for the deb.

File: schema/snapcraft.yaml [hunk 5]
```diff
@@ -73,13 +73,31 @@ properties:
...
+          a hyphen."
+      - pattern: "[^-]$"
+        validation-failure:
+          "{.instance!r} is not a valid snap name. Snap names cannot end with a
+          hyphen."
+      - not:
+          pattern: "--"
+        validation-failure:
+          "{.instance!r} is not a valid snap name. Snap names cannot have two
+          hyphens in a row."
```
Comment (elopio) [inline schema/snapcraft.yaml]: This diff shouldn't be here. What happened?

File: snap/snapcraft.yaml [hunk 6]
```diff
@@ -13,6 +13,7 @@ confinement: classic
 apps:
   snapcraft:
     command: bin/snapcraft
+    completer: snapcraft-completion
```
Comment (elopio) [inline snap/snapcraft.yaml]: This is also not related to this pr.

File: snapcraft/cli/_errors.py [hunk 7]
```diff
@@ -42,10 +54,15 @@ def exception_handler(exception_type, exception, exception_traceback, *,
...
 
-    if debug or not is_snapcraft_error:
+    if debug:
         traceback.print_exception(
             exception_type, exception, exception_traceback)
 
+    if not is_snapcraft_error:
+        echo.error('Sorry, Snapcraft had an internal error.')
+        if RavenClient is not None and _is_environment_ok():
+            _submit_trace(exception)
```
Comment (elopio) [inline snapcraft/cli/_errors.py]: When raven is not available, the message about sorry feels very empty. Shouldn't we give a link to report the issue, at least? Or point people to the forum to talk about it.

File: snapcraft/cli/_errors.py [hunk 8]
```diff
@@ -55,3 +72,20 @@ def exception_handler(exception_type, exception, exception_traceback, *,
             echo.error(str(exception))
 
     sys.exit(exit_code)
+
+
+def _is_environment_ok():
```
Comment (elopio) [inline snapcraft/cli/_errors.py]: I don't like the name of this function.
What about: `_send_error_data()`?

File: snapcraft/cli/_errors.py [hunk 9]
```diff
@@ -14,12 +14,24 @@
...
+import click
+# raven is not available on 16.04
+try:
+    from raven import Client as RavenClient
+    from raven.transport import RequestsHTTPTransport
+except ImportError:
+    RavenClient = None
+
+_MSG_INTERNAL_ERROR = 'Sorry, Snapcraft had an internal error.'
+_MSG_SEND_TO_SENTRY = 'To help us improve, would you like to send error data?'
```
Comment (elopio) [inline snapcraft/cli/_errors.py]: I would ask Mathew to improve the wording her.
It should mention that it's anonymous, that only the trace relevant to snapcraft and the snap being build will be collected, and that the info will never be shared outside of the snapcraft team. "error data" sounds scary to me.

File: snapcraft/cli/_errors.py (hunk 8 above)
Comment (sergiusens) [reply snapcraft/cli/_errors.py]: But it doesn't encompass everything

File: snap/snapcraft.yaml (hunk 6 above)
Comment (sergiusens) [reply snap/snapcraft.yaml]: It is because I rebased, but this PR is not against master, and even though the target branch is now updated, github missed out on that fact.

File: snapcraft/cli/_errors.py (hunk 7 above)
Comment (sergiusens) [reply snapcraft/cli/_errors.py]: This will only happen with the deb. This is according to the spec we already designed.

File: requirements.txt (hunk 4 above)
Comment (sergiusens) [reply requirements.txt]: In the PR now
